sycl/doc/design/KernelFusionJIT.md

```diff
@@ -0,0 +1,173 @@
+
+The detection at this stage is possible because even if the queue is in fusion mode, the individual kernels in the fusion list are added to the graph (and the leaves of memory records) such that other commands not part of the fusion process can correctly register them in their dependencies.
+
+Special case treatment and earlier detection during graph-construction in the `GraphBuilder` are necessary for the case where commands submitted to one queue have requirements or dependencies on commands submitted to another queue in fusion mode.
+
+If the fusion for a queue is aborted early, a warning can be printed to inform the user. This warning can be enabled by setting the environment variable `SYCL_RT_WARNING_LEVEL` to a value of  `1` or higher.
+
+### Event Handling
+
+The calls to `queue::submit` as well as to `queue::ext_codeplay_complete_fusion` need to return an event to allow synchronization with the execution of submitted kernels or the fused kernel.
+
+Returning the event associated with the individual kernels (`ExecCGCommand`) from `queue::submit` while in fusion mode would mean that these events become useless if the fusion mode is later on completed successfully because the individual kernels are removed from the graph and never executed in this case.
+
+To overcome this problem, a call to `queue::submit` while in fusion mode will instead return the event associated with the `KernelFusionCommand`.
+This event will always remain valid and allow for synchronization, independent of whether the fusion mode was terminated through early cancellation, `ext_codeplay_cancel_fusion`, or `ext_codeplay_complete_fusion`, as the KernelFusionCommand in all these scenarios has event dependencies on all relevant commands.
+
+Another important scenario is a call to `ext_codeplay_complete_fusion` after the fusion for this queue has been aborted early (see section on [Synchronization Behavior](#synchronization-behavior)).
+In this case, the individual kernels and the KernelFusionCommand could have been enqueued and even executed long ago, but the call to `complete_fusion` still needs to return a valid event for synchronization.
+To handle this case, the KernelFusionCommand for each queue remains in the fusion map until the next `ext_codeplay_start_fusion` on the same queue and does not undergo the cleanup process until its status has been set to `DELETE` by `ext_codeplay_start_fusion`.
+This way, the lifetime of the event associated with the KernelFusionCommand is extended such that it is still valid for synchronization when returned from a later call to `ext_codeplay_complete_fusion`.
```

Comment:
Does it mean that missing `ext_codeplay_complete_fusion` might lead to a memory leak?

Reply:
no, that's stored in a `unique_ptr`, so it gets cleaned up at some point (end of application in the worst case scenario)